### PR TITLE
Expose TF_RegisterFilesystemPlugin C API

### PR DESCRIPTION
**Pull request by yongtang**
This PR is part of the effort for modular file system.

In modular file system `RegisterFilesystemPlugin` was a C++ API and there is no directly way to call this API through python.

This PR exposes C API `TF_RegisterFilesystemPlugin`, so that it can be used in python bindings.

In addition, an experimental API `tf.experimental.register_filesystem_plugin` has been setup
so that it is possible to register a modular file system plugin with:
```
tf.experimental.register_filesystem_plugin(plugin_path)
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Review by mihaimaruseac — APPROVED** [review]
Thank you for this. I completely missed the Python part initially and had an item to work on this during the Windows filesystem conversion, but you got ahead of me :)

**Comment by mihaimaruseac** [comment]
This seems to have a lot of breakages, can you take a look please?

**Comment by yongtang** [comment]
Thanks @mihaimaruseac, I have updated the PR and all tests seem pass now. (I did an update to api_golden to cover the `tf.experimental.register_filesystem_plugin(plugin_path)`).

Please take a look and let me know if there are any other issues.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by yongtang** [comment]
@mihaimaruseac I noticed that GitHub is showing merge conflict. Do I need to rebase and update the PR?

**Comment by mihaimaruseac** [comment]
It should be good for now, it got imported internally and it seems most tests are passing. We'll need to wait for API owners approval, that would in ~24 hours.

**Comment by yongtang** [comment]
Thanks @mihaimaruseac! Let me know if there is anything else needs to be done.

**Comment by gbaned** [comment]
@yongtang  Can you please resolve conflicts? Thanks!

**Comment by yongtang** [comment]
@mihaimaruseac Wondering if there is any update on this PR or do you see if rebase and resolve the conflict is needed?

**Comment by mihaimaruseac** [comment]
Did not get approval yet from TF API owners, I think that due to team changes we will have to wait for the next week. Sorry for the delay.

There is no need to solve conflicts, I solved them internally already

**Pull request merged**
